typescript/lib/parsers/openai.ts

```diff
@@ -700,17 +720,32 @@ export class OpenAIChatModelParser extends ParameterizedModelParser<Chat.ChatCom
     const output = aiConfig.getLatestOutput(prompt);
     if (output != null) {
       if (output.output_type === "execute_result") {
-        // If the prompt has output saved, add it to the messages array
         if (output.metadata?.rawResponse?.role === "assistant") {
-          if (typeof output.data === "string") {
-            messages.push({
-              content: output.data,
-              // TODO (rossdanlm): Support function_call and don't rely on rawResponse
-              role: output.metadata?.rawResponse?.role,
-              function_call: output.metadata?.rawResponse?.function_call,
-              name: output.metadata?.rawResponse?.name,
-            });
+          output as ExecuteResult;
+          let content: string | null = null;
+          let function_call: FunctionData | undefined = undefined;
+
+          if (output.data?.hasOwnProperty("value")) {
+            const outputData = output.data as OutputDataWithValue;
+            if (typeof outputData.value === "string") {
+              content = outputData.value;
+            } else if (outputData.kind === "tool_calls") {
+              outputData as OutputDataWithToolCallsValue;
+              // Typescript schema does not support array of function calls yet,
+              // but Python does so doing this for forward compatibility
+              function_call =
+                outputData.value[outputData.value.length - 1].function;
+            }
+          } else if (typeof output.data === "string") {
+            content = output.data;
           }
+
+          messages.push({
+            content,
+            role: output.metadata?.rawResponse?.role,
+            ...(function_call !== undefined && { function_call }),
```

Comment:
We should update this to `tools` and `tool_choice` based on the latest OpenAI API schema, right? https://platform.openai.com/docs/api-reference/chat/create?lang=node.js

(Can come in a future diff)

Reply:
This is a good callout: `tool_calls` field is only used for assistant output type but that's fine because we're only setting this from assistant role anyways: https://github.com/openai/openai-node/blob/master/src/resources/chat/completions.ts#L147-L151
<img width="723" alt="Screenshot 2023-12-27 at 12 39 56" src="https://github.com/lastmile-ai/aiconfig/assets/151060367/95b681c6-1a4e-4061-822f-b64e5baaf44b">


Just a general note so you're aware, variable `message` is of a more general interface type `openai/resources/chat/ChatCompletionMessageParams` (defined here: https://github.com/openai/openai-node/blob/master/src/resources/chat/completions.ts#L445) which does not have the `function_call` or `tool_calls` field. Ex: here is the structure for user message: 
<img width="631" alt="Screenshot 2023-12-27 at 12 39 34" src="https://github.com/lastmile-ai/aiconfig/assets/151060367/954ff441-ff31-4ec9-bcef-977ef9853cc5">

Reply:
Yes should be fine. We can have a task for updating this, don't need to support tool and tool_calls in this diff

Reply:
Ok update, I tried doing this but couldn't import `ChatCompletionAssistantMessageParam`. COuld only work with the generalized `ChatCompletionMessageParam` object so just going to save it into the `function_call` field. Will leave comment with GH issue to track as backlog